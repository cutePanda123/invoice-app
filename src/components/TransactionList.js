import React from 'react';

const TransactionList = ({ items, onModifyItem, onDeleteItem }) => {
    return (
        <ul className = "list-group list-group-flush">
            {
                items.map((item) => (
                    <li className = "list-group-item d-flex justify-content-between align-items-center"
                        key = {item.id}
                    >
                        <span className="col-1 badge badge-primary">
                            {item.category.name}
                        </span>
                        <span className="col-5">{item.desc}</span>
                        <span className="col-2 font-weight-bold">
                            {(item.category.type === 'income') ? '+' : '-'}
                            ${item.amount}
                        </span>
                        <span className="col-2">{item.date}</span>
                        <button 
                            className="col-1 btn btn-primary"
                            onClick={() => {
                                onModifyItem(item);
                            }}
                        >edit</button>
                        <button 
                            className="col-1 btn btn-danger"
                            onClick={() => {
                                onDeleteItem(item);
                            }}
                        >delete</button>
                    </li>
                ))
            }
        </ul>
    );
}

export default TransactionList;