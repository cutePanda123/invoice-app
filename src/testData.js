export const testCategories = [
  {
    "name": "travel",
    "iconName": "ios-plane",
    "id": "1",
    "type": "outcome"
  },
  {
    "name": "food",
    "iconName": "ios-restaurant",
    "id": "2",
    "type": "outcome"
  },
  {
    "name": "shopping",
    "iconName": "ios-basket",
    "id": "3",
    "type": "outcome"
  },
  {
    "name": "electronic product",
    "iconName": "ios-phone-portrait",
    "id": "4",
    "type": "outcome"
  },
  {
    "name": "工资",
    "iconName": "ios-card",
    "id": "10",
    "type": "income"
  },
  {
    "name": "part-time job",
    "iconName": "ios-cash",
    "id": "11",
    "type": "income"
  },
  {
    "name": "investment",
    "iconName": "logo-yen",
    "id": "12",
    "type": "income"
  }
];
  
export const testItems = [
  {
    "id": 0,
    "desc": "travel in Seattle",
    "amount": 2000,
    "date": "2020-10-20",
    "categoryId": 1
  },
  {
    "id": 1,
    "desc": "travel in San Jose",
    "amount": 4000,
    "date": "2020-12-10",
    "categoryId": 1
  },
  {
    "id": 2,
    "desc": "stock investement",
    "amount": 14000,
    "date": "2020-12-15",
    "categoryId": 2
  }
];

export const fakeTransaction = {
  "id": 0,
  "desc": "fake new transaction",
  "amount": 14000,
  "date": "2020-12-10",
  "categoryId": 2
};