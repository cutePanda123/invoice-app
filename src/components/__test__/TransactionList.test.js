import React from 'react';
import { shallow } from 'enzyme';
import Ionicon from 'react-ionicons';
import TransactionList from '../TransactionList';
import { testItems as items, testCategories as categories } from '../../testData';

const itemWithCategories = items.map((item) => {
    item.category = categories[item.categoryId];
    return item;
});

const props = {
    items: itemWithCategories,
    onModifyItem: jest.fn(),
    onDeleteItem: jest.fn()
};

let wrapper = null;

describe('TransactionList test', () => {
    beforeEach(() => {
        wrapper = shallow(<TransactionList {...props} />);
    });

    it('snapshot test', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should render correct number of items', () => {
        expect(wrapper.find('.list-group-item').length).toEqual(itemWithCategories.length);
    });

    it('should render correct icons for each item', () => {
        const icons = wrapper.find('.list-group-item').first().find(Ionicon);
        expect(icons.length).toEqual(3);
        expect(icons.first().props().icon)
            .toEqual(itemWithCategories[0].category.iconName);
    });

    it('should trigger correct callback', () => {
        const firstItem = wrapper.find('.list-group-item').first();
        firstItem.find('a').first().simulate('click');
        expect(props.onModifyItem).toHaveBeenCalledWith(itemWithCategories[0]);

        firstItem.find('a').last().simulate('click');
        expect(props.onDeleteItem).toHaveBeenCalledWith(itemWithCategories[0]);
    });
});

