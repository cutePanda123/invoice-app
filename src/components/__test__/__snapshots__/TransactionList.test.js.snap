// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionList test snapshot test 1`] = `
<ul
  className="list-group list-group-flush"
>
  <li
    className="list-group-item d-flex justify-content-between align-items-center"
    key="0"
  >
    <span
      className="col-1"
    >
      <Ionicon
        beat={false}
        className="rounded-circle"
        color="#fff"
        fontSize="30px"
        icon="ios-plane"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#007bff",
            "padding": "5px",
          }
        }
      />
    </span>
    <span
      className="col-5"
    >
      travel in Seattle
    </span>
    <span
      className="col-2 font-weight-bold"
    >
      -
      $
      2000
    </span>
    <span
      className="col-2"
    >
      2020-10-20
    </span>
    <a
      className="col-1"
      onClick={[Function]}
    >
      <Ionicon
        beat={false}
        className="rounded-circle"
        color="#fff"
        fontSize="30px"
        icon="ios-create-outline"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#28a745",
            "padding": "5px",
          }
        }
      />
    </a>
    <a
      className="col-1"
      onClick={[Function]}
    >
      <Ionicon
        beat={false}
        className="rounded-circle"
        color="#fff"
        fontSize="30px"
        icon="ios-close"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#dc3545",
            "padding": "5px",
          }
        }
      />
    </a>
  </li>
  <li
    className="list-group-item d-flex justify-content-between align-items-center"
    key="1"
  >
    <span
      className="col-1"
    >
      <Ionicon
        beat={false}
        className="rounded-circle"
        color="#fff"
        fontSize="30px"
        icon="ios-plane"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#007bff",
            "padding": "5px",
          }
        }
      />
    </span>
    <span
      className="col-5"
    >
      travel in San Jose
    </span>
    <span
      className="col-2 font-weight-bold"
    >
      -
      $
      4000
    </span>
    <span
      className="col-2"
    >
      2020-12-10
    </span>
    <a
      className="col-1"
      onClick={[Function]}
    >
      <Ionicon
        beat={false}
        className="rounded-circle"
        color="#fff"
        fontSize="30px"
        icon="ios-create-outline"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#28a745",
            "padding": "5px",
          }
        }
      />
    </a>
    <a
      className="col-1"
      onClick={[Function]}
    >
      <Ionicon
        beat={false}
        className="rounded-circle"
        color="#fff"
        fontSize="30px"
        icon="ios-close"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#dc3545",
            "padding": "5px",
          }
        }
      />
    </a>
  </li>
  <li
    className="list-group-item d-flex justify-content-between align-items-center"
    key="2"
  >
    <span
      className="col-1"
    >
      <Ionicon
        beat={false}
        className="rounded-circle"
        color="#fff"
        fontSize="30px"
        icon="logo-yen"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#007bff",
            "padding": "5px",
          }
        }
      />
    </span>
    <span
      className="col-5"
    >
      stock investement
    </span>
    <span
      className="col-2 font-weight-bold"
    >
      +
      $
      14000
    </span>
    <span
      className="col-2"
    >
      2020-12-15
    </span>
    <a
      className="col-1"
      onClick={[Function]}
    >
      <Ionicon
        beat={false}
        className="rounded-circle"
        color="#fff"
        fontSize="30px"
        icon="ios-create-outline"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#28a745",
            "padding": "5px",
          }
        }
      />
    </a>
    <a
      className="col-1"
      onClick={[Function]}
    >
      <Ionicon
        beat={false}
        className="rounded-circle"
        color="#fff"
        fontSize="30px"
        icon="ios-close"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#dc3545",
            "padding": "5px",
          }
        }
      />
    </a>
  </li>
</ul>
`;
